dask/array/creation.py

```diff
@@ -501,6 +502,82 @@ def diag(v):
     return Array(graph, name, (chunks_1d, chunks_1d), dtype=v.dtype)
 
 
+@wraps(np.diagonal)
+def diagonal(a, offset=0, axis1=0, axis2=1):
+    name = 'diagonal-' + tokenize(a, offset, axis1, axis2)
+
+    if a.ndim < 2:
+        # NumPy uses `diag` as we do here.
+        raise ValueError("diag requires an array of at least two dimensions")
+
+    def _axis_fmt(axis, name, ndim):
+        if axis < 0:
+            t = ndim + axis
+            if t < 0:
+                msg = "{}: axis {} is out of bounds for array of dimension {}"
+                raise AxisError(msg.format(name, axis, ndim))
+            axis = t
+        return axis
+
+    axis1 = _axis_fmt(axis1, "axis1", a.ndim)
+    axis2 = _axis_fmt(axis2, "axis2", a.ndim)
+
+    if axis1 == axis2:
+        raise ValueError("axis1 and axis2 cannot be the same")
+
+    if isinstance(a, np.ndarray):
+        return diagonal(asarray(a), offset, axis1, axis2)
+
+    if not isinstance(a, Array):
+        raise TypeError("a must be a dask array or numpy array, "
+                        "got {0}".format(type(a)))
```

Comment:
Would suggest cutting both of these `if`s and just having `a = asarray(a)`. This will convert the NumPy array to a Dask Array. Also it will raise appropriately if a Dask Array cannot be created (or is not already present).